### PR TITLE
fix: exclude .apm-pin from skill deploy copytree

### LINKED ISSUE #1150: [BUG] .apm-pin marker leaks into skill deploy targets on subsequent installs
## Description

When a skill package is installed from a remote dependency, the `.apm-pin` cache marker (introduced in #1137) is correctly written into the `apm_modules/` cache directory during the lockfile phase. However, on the **second and subsequent** `apm install` runs, the marker file gets copied into the skill's deploy target (e.g., `.agents/skills/<skill-name>/.apm-pin`) because `shutil.copytree` in the skill integrator doesn't filter it out.

## Steps to reproduce

1. Create a project with a remote skill dependency (e.g., a `claude_skill` package type)
2. Run `apm install` — first run is clean
3. Run `apm install` again — `.apm-pin` now appears in the deploy target

## Expected behavior

The `.apm-pin` file should only exist under `apm_modules/` (which is gitignored). It should not be copied into deploy targets like `.agents/skills/`.

## Actual behavior

After the second install, `.apm-pin` appears in `.agents/skills/<skill-name>/` alongside `SKILL.md`. Since `.agents/skills/` is typically committed to source control, this shows up as an untracked/dirty file in git.

## Root cause

The install pipeline phase order is: resolve → download → **integrate** → cleanup → **lockfile**

- The **lockfile phase** writes `.apm-pin` into `apm_modules/<owner>/<repo>/<path>/` (correct)
- On subsequent runs, the **integrate phase** calls `shutil.copytree(source_path, skill_dir, ignore=ignore_symlinks)` at `skill_integrator.py:392`
- The `ignore_symlinks` callback only filters symlinks, not `.apm-pin`
- Since the marker already exists from the previous run's lockfile phase, it gets swept up in the copy

## Suggested fix

Add `.apm-pin` to the ignore function used by `shutil.copytree` in `skill_integrator.py`. Something like:

```python
def _ignore_non_content(directory, contents):
    ignored = set()
    for name in contents:
        path = os.path.join(directory, name)
        if os.path.islink(path) or name == ".apm-pin":
            ignored.add(name)
    return ignored

shutil.copytree(source_path, skill_dir, ignore=_ignore_non_content)
```

The same pattern may apply to the sub-skill `copytree` call at line ~649.

## Context

This is a small interaction between the cache-pin feature (#1137, #1142) and the existing skill integration copy logic. The pin feature itself is working correctly — it just needs to be excluded from the deploy copy like it's already excluded from content hashing in `_EXCLUDED_ROOT_FILES`.

## Environment

- apm 0.12.2
- macOS (Darwin 25.4.0)

### PR DESCRIPTION
**Pull request by tillig**
## Summary

- Adds `ignore_non_content` copytree callback to `security/gate.py` that filters both symlinks and `.apm-pin` cache markers
- Applies it to all skill-deploy copytree paths: skill integrator (3 sites), bundle packer, bundle unpacker, and plugin parser skill copies
- Prevents the `.apm-pin` cache marker (written to `apm_modules/` by the lockfile phase) from leaking into deploy targets like `.agents/skills/` on subsequent installs

Fixes #1150 (regression from #1137)

## Root cause

The install pipeline runs phases in this order: resolve → download → **integrate** → cleanup → **lockfile**. The lockfile phase writes `.apm-pin` markers into `apm_modules/` cache directories. On subsequent installs, the integrate phase's `shutil.copytree` copies the entire cache directory — including the now-present `.apm-pin` — into deploy targets. The `ignore=ignore_symlinks` callback only filtered symlinks, not the pin marker.

## Test plan

- [x] New unit tests in `test_ignore_non_content.py` (6 cases): root filtering, nested filtering, symlink filtering preserved, normal files pass through, end-to-end apm_modules simulation, similar-name boundary case
- [x] Existing `test_symlink_containment.py` source-level guard updated and passing
- [x] Full unit suite passes (7624 passed, 4 pre-existing env-specific failures unrelated to this change)
- [x] Reproduced the bug locally before fix, confirmed fix resolves it

### CODE REVIEW

File: src/apm_cli/security/gate.py [hunk 1]
```diff
@@ -210,3 +210,20 @@ def _build_verdict(
...
+
+    Excludes symlinks (security) and the ``.apm-pin`` cache marker, which
+    belongs exclusively in ``apm_modules/`` and must not leak into deploy
+    targets when skills are copied out.
+    """
+    return [
+        c
+        for c in contents
+        if (Path(directory) / c).is_symlink() or c == _APM_PIN_FILENAME
+    ]
```
[the remark this answers is not in the document]
Comment (tillig) [reply src/apm_cli/security/gate.py]: Unable to reuse the constant because it would cause a circular dependency. Added comment to that effect.

File: tests/unit/test_ignore_non_content.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        src = tmp_path / "src"
+        src.mkdir()
+        (src / "SKILL.md").write_text("# My Skill", encoding="utf-8")
+        target = tmp_path / "secret.txt"
+        target.write_text("secret", encoding="utf-8")
+        (src / "evil-link").symlink_to(target)
+
+        dest = tmp_path / "dest"
+        shutil.copytree(src, dest, ignore=ignore_non_content)
+
```
[the remark this answers is not in the document]
Comment (tillig) [reply tests/unit/test_ignore_non_content.py:50]: Added a try/catch with a skip so the test will be skipped if symlinks aren't supported.

**Comment by danielmeppiel** [comment]
## Panel follow-ups landed (commit `306647c9`)

Addressing the prior advisory panel's recommended follow-ups directly on this branch (maintainer-can-modify is on). Re-ran a focused 3-panelist review (python-architect for cache invariants, supply-chain-security for the gate.py change, test-coverage-expert for test adequacy) on the new diff.

### What changed

| File | Change |
|---|---|
| `src/apm_cli/security/gate.py` | Replaced hardcoded `_APM_PIN_FILENAME` (with stale circular-import comment) with lazy `from apm_cli.install.cache_pin import MARKER_FILENAME`. Single source of truth. |
| `src/apm_cli/utils/content_hash.py` | Imports `MARKER_FILENAME` from `cache_pin` instead of hardcoding `".apm-pin"`. |
| `src/apm_cli/deps/plugin_parser.py` | Extended `ignore_non_content` guard symmetrically to agents (lines 463/465) and hooks (lines 553/555) copytree sites; removed dead `_ignore_symlinks` local helper. All 7 plugin_parser copytree sites now use the canonical guard. |
| `tests/integration/test_skill_install.py` | Added `test_reinstall_does_not_leak_apm_pin_to_deploy_targets` (reuses existing `temp_project` + `apm_command` fixtures, no fixture sprawl). |
| `tests/unit/test_symlink_containment.py` | Added `TestIgnoreNonContentSourceGuard` parametrized over `plugin_parser` (min_refs=7), `packer` (2), `unpacker` (2) -- catches future copytree sites that drop the guard. |
| `CHANGELOG.md` | Dropped bold from `.apm-pin` Fixed entry to match siblings (#1147, #1158). |

### Cache safety verified

The cache-architect persona traced the full `.apm-pin` lifecycle and confirmed:
- `.apm-pin` is written ONLY by `cache_pin.sync_markers_for_lockfile()` into `apm_modules/<owner>/<repo>/`; never expected under deploy targets.
- The "circular import" comment that was on `_APM_PIN_FILENAME` is false -- `cache_pin.py` imports only stdlib (`json`, `pathlib`, `typing`) plus a lazy console helper. Zero security imports.
- `ignore_non_content` is a strict superset of the removed `_ignore_symlinks`. Extending it to agents/hooks copytree is pure defense-in-depth (those subdirs never legitimately contain `.apm-pin`).
- All 6 deploy sites are self-healing (`rmtree` then `copytree`), so the leak only ever appears on first install of an already-cached package.

### Validation

- Lint clean: `ruff check src/ tests/` and `ruff format --check src/ tests/` both silent.
- 7692 unit tests pass + 8 `test_skill_install.py` integration tests pass (incl. the new regression test).
- Source-guard subtests pass for all 3 modules.

### Panel ship recommendation

Three panelists ran in parallel. Architect's actionable finding (bump `packer`/`unpacker` `min_refs` from 1 to 2 for tighter slack) was applied. Security clean (one nit on lazy-vs-eager import consistency, set aside for symmetry with packer/unpacker pattern). Test-coverage flagged adding integration tests for agent/hook plugin types end-to-end (recommended, not blocking) -- set aside per anti-fixture-sprawl directive since the source-guard with `min_refs=7` proves all 7 plugin_parser copytree sites use the guard.

Co-authored with @copilot.

**Review by danielmeppiel — APPROVED** [review]
